File: solutions/java/relative-distance/1/src/main/java/RelativeDistance.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class RelativeDistance {
    private final Map<String, List<String>> familyGraph;
    RelativeDistance(Map<String, List<String>> familyTree) {
        Map<String, List<String>> graph = new HashMap<>();
        familyTree.forEach((k, v) -> {
            graph.putIfAbsent(k, new ArrayList<>());
            v.forEach(e -> {
                graph.get(k).add(e);
                graph.putIfAbsent(e, new ArrayList<>());
                graph.get(e).add(k);
                v.stream().filter(sibling -> !sibling.equals(e))
                    .forEach(sibling -> graph.get(e).add(sibling));
            });
        });
        this.familyGraph = graph;
    }

    int degreeOfSeparation(String personA, String personB) {
        var visited = new HashSet<String>();
        var queue = new ArrayDeque<Pair>();
        queue.add(new Pair(personA, 0));
        visited.add(personA);
        while (!queue.isEmpty()) {
            var cur = queue.remove();
            var curName = cur.name();
            if (curName.equals(personB)) {
                return cur.degree();
            }
            var adjacent = familyGraph.get(curName);
            int nextDegree = cur.degree() + 1;
            for (var name: adjacent) {
                if (visited.contains(name)) {
                    continue;
                }
                if (name.equals(personB)) {
                    return nextDegree;
                }
                queue.add(new Pair(name, nextDegree));
                visited.add(name);
            }
        }
        return -1;
    }

    record Pair(String name, int degree) {}
}

File: solutions/javascript/relative-distance/1/relative-distance.js
export const degreesOfSeparation = (familyTree, personA, personB) => {
  const graph = makeFamilyGraph(familyTree)
  const queue = []
  const visited = new Set()
  queue.push({name: personA, degree: 0})
  visited.add(personA)
  while (queue.length > 0) {
    const cur = queue.shift()
    const curName = cur.name
    if (curName === personB) {
      return cur.degree
    }
    const nextDegree = cur.degree + 1
    for (const adjacent of graph.get(curName)) {
      if (adjacent === personB) {
        return nextDegree
      }
      if (visited.has(adjacent)) {
        continue
      }
      queue.push({name: adjacent, degree: nextDegree})
      visited.add(adjacent)
    }
  }
  return -1
}

function makeFamilyGraph(familyTree) {
  const graph = new Map()
  for (const [parent, children] of Object.entries(familyTree)) {
    if (!graph.has(parent)) {
      graph.set(parent, new Set())
    }
    const parentSet = graph.get(parent)
    children.forEach(child => {
      parentSet.add(child)
    })
    children.forEach(child => {
      if (!graph.has(child)) {
        graph.set(child, new Set())
      }
      const childSet = graph.get(child)
      childSet.add(parent)
      children.forEach(sibling => {
        childSet.add(sibling)
      })
    })
  }
  return graph
}